// babel-plugin-inline-react-svg/default-props.js

> babel-plugin-inline-react-svg@0.4.0 sanity /Users/joe_lencioni/babel-plugin-inline-react-svg
> babel-node test/sanity.js

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MyClassIcon = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

exports.MyFunctionIcon = MyFunctionIcon;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var MySvg = function () {
  function MySvg(props) {
    return _react2['default'].createElement(
      'svg',
      props,
      _react2['default'].createElement('title', null),
      _react2['default'].createElement('circle', {
        cx: '1038.5',
        cy: '467.01',
        r: '72.28',
        transform: 'rotate(-45 84.888 1433.508)',
        fill: '#da2244',
        stroke: '#f2f2f2',
        strokeLinecap: 'round',
        strokeLinejoin: 'round',
        strokeWidth: '7'
      }),
      _react2['default'].createElement('path', {
        fill: '#da2244',
        stroke: '#f2f2f2',
        strokeLinecap: 'round',
        strokeLinejoin: 'round',
        strokeWidth: '7',
        d: 'M47.57 103.99l56.42-56.42M45.8 45.87l59.9 59.9'
      }),
      _react2['default'].createElement(
        'text',
        null,
        'This is an opening ',
        '{',
        ' and this is a closing ',
        '}',
        '.'
      ),
      _react2['default'].createElement(
        'style',
        null,
        'text',
        '{',
        'color:#00f',
        '}'
      )
    );
  }

  return MySvg;
}();

MySvg.defaultProps = {
  dataName: 'Livello 1',
  id: 'Livello_1',
  viewBox: '0 0 151.57 151.57',
  xmlns: 'http://www.w3.org/2000/svg'
};
function MyFunctionIcon() {
  return _react2['default'].createElement(MySvg, null);
}

var MyClassIcon = exports.MyClassIcon = function (_React$Component) {
  _inherits(MyClassIcon, _React$Component);

  function MyClassIcon() {
    _classCallCheck(this, MyClassIcon);

    return _possibleConstructorReturn(this, (MyClassIcon.__proto__ || Object.getPrototypeOf(MyClassIcon)).apply(this, arguments));
  }

  _createClass(MyClassIcon, [{
    key: 'render',
    value: function () {
      function render() {
        return _react2['default'].createElement(MySvg, null);
      }

      return render;
    }()
  }]);

  return MyClassIcon;
}(_react2['default'].Component);
